src/flow.ts

```diff
@@ -238,6 +236,11 @@ export class Flow {
   localFlags: LocalFlags[] = [];
   /** Field flags on `this`. Constructors only. */
   thisFieldFlags: Map<Field,FieldFlags> | null = null;
+  /** type narrow */
+  narrowedTypes: NarrowedTypeMap | null = null;
+  /** conditional type narrow, eg if (condi) */
+  conditionalNarrowedType: TypeNarrowChecker = new TypeNarrowChecker();
```

Comment:
Hnnm, can you explain why this additional map is necessary? It seems to me that anything that is conditional would be a `narrowedType` limited to the respective flow it is narrowed in. Say there is an `if (cond)`, then the inner flow within the if would the state after `cond` as `narrowedTypes`, and at the end of the if, the inner flow would be merged with the outer one again, updating the successor flow's `narrowedTypes`. As such I don't quite see what `conditionoalNarrowedTypes` does, as within a flow, there wouldn't be conditionals?

Reply:
The name may cause some misunderstandings. This is conditional type narrowing checker which will collect the info of type narrowing and decide how type narrowed in then and else branch according to the condition expression. 
Do you think renaming it as `conditionalTypeNarrowChecker` is more straightforward?

Reply:
I see, one is a map and the other is a checker object. Would it be possible to abstract the checker object away? Looks like there's always a `new` one assigned to a `Flow` anyway, so the flow could include the respective methods directly perhaps.

Reply:
> Would it be possible to abstract the checker object away?
Abstract it away is difficult because map relies on each flow but check is for the flow and the forked flow. Actually it should not create a new checker for each flow. I will move it into constructor and inherit it during fork.